README.md
# Automate Let's Encrypt SSL/TLS Certificate Deployment to NameSilo

### docker-compose.yml
```yaml
version: '3'
services:
  app:
    container_name: namesilo_letsencrypt_app
    image: elquimista/namesilo-letsencrypt:latest
    stdin_open: true
    tty: true
    restart: unless-stopped
    volumes:
      - ./config.py:/app/config.py
      - ./letsencrypt:/etc/letsencrypt
```

### Start
```sh
docker-compose up -d
```

### Generate certificates for the first time
```sh
docker-compose exec app sh
./namesilo-certbot.sh --email me@blue.planet -d "example.com"
```
This command will generate certificate key files under `letsencrypt` folder (specified in the docker compose volume section). The script will take 60 minutes to finish execution (due to Namesilo's DNS propagation taking approximately 60 minutes at the time of writing this. Feel free to adjust the waiting minutes appropriately in `config.py`).

### Using certificates
Now you have certificate files generated inside `letsencrypt` folder, you can use them by mounting them to another Docker container that runs Nginx or symlink them if you run Nginx directly on the host OS. It's up to you.

### Renewal of certificates
Certificates are renewed automatically with certbot renew command and cronjob. It's baked into this Docker image so you don't have to do anything.

---

Credit to [ethauvin/namesilo-letsencrypt](https://github.com/ethauvin/namesilo-letsencrypt) for the under the hood heavylifting.


namesilo-letsencrypt-0.9.2/CHANGELOG.md
# Changelog

## [0.9.2](https://github.com/ethauvin/namesilo-letsencrypt/tree/0.9.2) (2020-05-28)

[Full Changelog](https://github.com/ethauvin/namesilo-letsencrypt/compare/0.9.1...0.9.2)

**Closed issues:**

- Do not support multiple \_acme-challenge records [\#1](https://github.com/ethauvin/namesilo-letsencrypt/issues/1)

**Merged pull requests:**

- make python3.4 compatible. add subdomains support [\#2](https://github.com/ethauvin/namesilo-letsencrypt/pull/2) ([Junker](https://github.com/Junker))

## [0.9.1](https://github.com/ethauvin/namesilo-letsencrypt/tree/0.9.1) (2020-03-15)

[Full Changelog](https://github.com/ethauvin/namesilo-letsencrypt/compare/0.9.0...0.9.1)

## [0.9.0](https://github.com/ethauvin/namesilo-letsencrypt/tree/0.9.0) (2019-04-30)

[Full Changelog](https://github.com/ethauvin/namesilo-letsencrypt/compare/8a91146faebd196b223f7653e470428130f26246...0.9.0)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


namesilo-letsencrypt-0.9.2/cleanup.py
#!/usr/bin/env python3

#  cleanup.py
#
#  Copyright (c) 2019-2020, Erik C. Thauvin (erik@thauvin.net)
#  All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are met:
#
#    Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
#
#    Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
#
#    Neither the name of this project nor the names of its contributors may be
#    used to endorse or promote products derived from this software without
#    specific prior written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
#  THE POSSIBILITY OF SUCH DAMAGE.

import os
import sys
import tempfile
import urllib.request

import tldextract
import untangle

from config import apikey

domain = os.environ['CERTBOT_DOMAIN']
tmpdir = os.path.join(tempfile.gettempdir(), "CERTBOT_" + domain)

if "NAMESILO_API" in os.environ:
    apikey = os.environ['NAMESILO_API']

tld = tldextract.extract(domain)
nsdomain = tld.domain + "." + tld.suffix

url = "https://www.namesilo.com/api/dnsDeleteRecord\
?version=1&type=xml&key=" + apikey + "&domain=" + nsdomain + "&rrid="


def getrequest(record_id):
    return urllib.request.Request(
        url + record_id,
        data=None,
        headers={
            'User-Agent': ('Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:74.0) '
                           'Gecko/20100101 Firefox/74.0')
        }
    )


idFile = os.path.join(tmpdir, "RECORD_ID")
if os.path.isfile(idFile):
    f = open(idFile, "r")
    for line in f:
        with urllib.request.urlopen(getrequest(line.rstrip())) as response:
            html = response.read()
        xml = untangle.parse(str(html, 'utf-8'))
        if xml.namesilo.reply.code.cdata != '300':
            print("{}: {} ({})".format(
                domain,
                xml.namesilo.reply.detail.cdata,
                xml.namesilo.reply.code.cdata), file=sys.stderr)
            if xml.namesilo.reply.code.cdata != '280':
                sys.exit(1)
    f.close()
    os.remove(idFile)


namesilo-letsencrypt-0.9.2/authenticator.py
#!/usr/bin/env python3

#  authenticator.py
#
#  Copyright (c) 2019-2020, Erik C. Thauvin (erik@thauvin.net)
#  All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are met:
#
#    Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
#
#    Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
#
#    Neither the name of this project nor the names of its contributors may be
#    used to endorse or promote products derived from this software without
#    specific prior written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
#  THE POSSIBILITY OF SUCH DAMAGE.

import os
import sys
import tempfile
import time
import urllib.request

import tldextract
import untangle

from config import apikey, wait


def sleep(minutes):
    if minutes < 16:
        minutes = 16
    time.sleep(minutes * 60)


domain = os.environ['CERTBOT_DOMAIN']
validation = os.environ['CERTBOT_VALIDATION']
tmpdir = os.path.join(tempfile.gettempdir(), "CERTBOT_" + domain)
rrhost = "_acme-challenge"

if "NAMESILO_API" in os.environ:
    apikey = os.environ['NAMESILO_API']

tld = tldextract.extract(domain)
nsdomain = tld.domain + "." + tld.suffix
if tld.subdomain:
    rrhost += "." + tld.subdomain

url = "https://www.namesilo.com/api/dnsAddRecord?\
version=1&type=xml&key=" + apikey + "&domain=" + nsdomain + "&rrtype=TXT\
&rrhost=" + rrhost + "&rrvalue=" + validation + "&rrttl=3600"

req = urllib.request.Request(
    url,
    data=None,
    headers={
        'User-Agent': ('Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:74.0) '
                       'Gecko/20100101 Firefox/74.0')
    }
)

with urllib.request.urlopen(req) as response:
    html = response.read()
xml = untangle.parse(str(html, 'utf-8'))

if not os.path.exists(tmpdir):
    os.mkdir(tmpdir, 0o700)

if xml.namesilo.reply.code.cdata == '300':
    f = open(os.path.join(tmpdir, "RECORD_ID"), "a+")
    print(xml.namesilo.reply.record_id.cdata, file=f)
    f.close()
else:
    print("{}: {} ({})".format(domain,
                               xml.namesilo.reply.detail.cdata,
                               xml.namesilo.reply.code.cdata), file=sys.stderr)
    sys.exit(1)

# Sleep X minutes
sleep(wait)
